### ISSUE TITLE
.net 8, HttpClient, adding ssl cert after first use, the cert does not get used 

### ISSUE THREAD

**Issue by chacham2** [dotnet/runtime#96494]
### Description

I do not know how to verify this behavior other than the one server (webhooks) i am connecting to. But, based on my experience, it seems that in .net 8, If an HttpClient uses .GetAsync() or .PostAsync(), and then adds an ssl certificate to the handler, the cert does not get used for following calls. The same worked in .net 7.

### Reproduction Steps

I do not know how to show if an ssl certificate is being used. The idea is simple, add the ssl certificate after the first request:

	Imports System.Net.Http
	Imports System.Security.Cryptography.X509Certificates

	Public Class Form1
		Private Async Sub Form1_Load(Sender As Object, Arguments As EventArgs) Handles MyBase.Load
			Dim Http_Client_Handler As New HttpClientHandler
			Dim Http_Client As New HttpClient(Http_Client_Handler)

			Using Response As HttpResponseMessage = Await Http_Client.GetAsync($"https://example.com")
				Debug.WriteLine($"Result: {Await Response.Content.ReadAsStringAsync}")
				Debug.WriteLine($"Status: {Response.StatusCode}")
			End Using

			' Add SSL Certificate
			Using X509_Store As New X509Store(StoreLocation.LocalMachine)
				X509_Store.Open(OpenFlags.ReadOnly)
				Http_Client_Handler.ClientCertificates.Add(X509_Store.Certificates.Find(X509FindType.FindBySubjectDistinguishedName, $"CN={ ...}", True)(0))
			End Using

			Using Response As HttpResponseMessage = Await Http_Client.GetAsync($"https://example.com")
				Debug.WriteLine($"Result: {Await Response.Content.ReadAsStringAsync}")
				Debug.WriteLine($"Status: {Response.StatusCode}")
			End Using
		End Sub
	End Class

### Expected behavior

The second connection should work without issue.

### Actual behavior

The second request returns a 400, with the error that no ssl certificate was used.

### Regression?

This worked in .net 7.

### Known Workarounds

Add the ssl certificate first, or use a second HttpClient.

### Configuration

Windows 10, .net 8, x64.

**Comment by wfurt** [comment]
I don't think it was intention ... perhaps just luck. Perhaps somebody else will have some more thoughts on this.

**Comment by CarnaViire** [comment]
@chacham2 I'm trying to fully understand the scenario....

The first request is being done _without any certs_, and it works, but the second one does not? Why does the server allow the first request without certs, but not the second? Are the calls different?

Or, does the first one actually _fail_? So the behavior you expect:
`call fails without cert -> you add the cert -> the next call works`
but the one you get is:
`call fails without cert -> you add the cert -> the next still fails`
?

**Comment by wfurt** [comment]
simple complete repro would be the best @chacham2

**Comment by CarnaViire** [comment]
Triage: We need to investigate in 9.0

**Comment by rzikm** [comment]
I have a hunch this is caused by https://github.com/dotnet/runtime/pull/88214

<img width="978" alt="image" src="https://github.com/dotnet/runtime/assets/32671551/f7d6dd6b-9575-4a64-853a-e6a8b12d4574">

**Comment by wfurt** [comment]
it may be. `_clientCertificateOptions == ClientCertificateOption.Manual` -> the should not be executed with `Automatic` as well as selection callback should still work.

**Comment by chacham2** [comment]
@wfurt 
> simple complete repro would be the best @chacham2

I didn't provide a working repro because i'd have to share user id and password. Instead, i got it down to a repeatable proof of concept which i posted in the first post, but i changed the server name and remove private information. I wasn't sure what else to do.

**Comment by YePiaoRan17** [comment]
I have similar problems.
![屏幕截图 2024-01-23 185348](https://github.com/dotnet/runtime/assets/42258736/2177a498-4063-4624-9260-63898c602e00)

**Comment by wfurt** [comment]
Did you try the selection callback @YePiaoRan17 ? If the root cause really is https://github.com/dotnet/runtime/pull/88214 I don't think we want to go back. And modifying the collection while in use is on safe in general.

**Comment by rzikm** [comment]
> And modifying the collection while in use is on safe in general.

This. The HttpClient is intended to be configured before you send the first request. Many properties will throw if you attempt to modify them, but collections are difficult because when the getter is called, we don't know if you are going to modify it or not.
